### PR TITLE
Some updates to the HomeBuilder

### PR DESCRIPTION
**Pull request by martinvw**
New features

- Added semantic labels to tags Floors, Rooms and objects
- Added a Cellar as a Floor
- Added support for more languages which where already available as translation

Changes to the build:

- Integrated the webpack build inside Maven
- Only allow minor updates in the package.json to make sure new version will not easily break the build, if semver fails again
- Remove dist folder because it should be generated through the build

Small updates because of some updated dependencies

- Switched from 'hidden-xs-up' to 'd-none' as proposed in bootstrap migration guide
- Switched from 'scope' to 'slot-scope' as suggested by vue 2.5 migration
- Added some additional parents in the vue schema because they were needed

Small fixes:

- removed osgi import from manifest

[org.openhab.ui.homebuilder-2.4.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-core/files/2502545/org.openhab.ui.homebuilder-2.4.0-SNAPSHOT.jar.zip)

### CODE REVIEW

**Comment by ghys** [comment]
@martinvw
FYI this is the file with the semantic tags:
https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.semantics/model/SemanticTags.csv

and these are the property files with the list of translated terms & synonyms:
https://github.com/eclipse/smarthome/tree/master/bundles/core/org.eclipse.smarthome.core.semantics/src/main/resources

**Comment by martinvw** [comment]
> and these are the property files with the list of translated terms & synonyms:

So I just copy the en to nl and translate that?

**Comment by ghys** [comment]
> So I just copy the en to nl and translate that?

Yes - off-topic now, but if your intent is to add NL support for HABot you will also need to add stuff here: https://github.com/openhab/org.openhab.ui.habot/tree/master/src/main/resources
- `answers_nl.properties`
- `colors_pl.properties`
- ("tagattributes" are not needed, these files are not used anymore with the semantics)
- a `nl` folder under `train` with the training data (use the other languages as examples)

**Comment by martinvw** [comment]
@kubawolanin @ghys It would be great if one of you or even better of you could do a review. @ghys I understood that you also have experience with this technology

File: bundles/org.openhab.ui.homebuilder/web/src/restItems.js [hunk 1]
```diff
@@ -4,19 +4,15 @@ import { floors, objects, OBJECTS_SUFFIX } from './definitions'
...
-
-    return items;
+    return [{
+        type: 'Group',
+        name: floor.abbr,
+        label: floor.name || floor.value,
+        category: model.itemsIcons ? floor.icon : '',
+        groupNames: ['Home'],
+        entryType: 'floor',
+        tags: floor.tags
```
Comment (kubawolanin) [inline bundles/org.openhab.ui.homebuilder/web/src/restItems.js]: We should make sure that `floor.tags` (as well as `room.tags` and `object.tags`) are not `undefined`.
E.g. 
```
tags: floor.tags || []
```
Or create a reusable method, e.g. `getTags` that'd handle empty state:
```
function getTags(entry) {
    return entry.tags || [];
}
```

And then:
```
tags: getTags(floor)
...
tags: getTags(object)
```
and so on.

Right now the set of definitions is finite and static, but in the future we may handle it dynamically.

File: bundles/org.openhab.ui.homebuilder/web/src/restItems.js [hunk 2]
```diff
@@ -65,7 +62,7 @@ export function addObjects(room, model, floorPrefix, roomObjects) {
             floorPrefix + room.value,
             GROUP_PREFIX + object.value
         ],
-        tags: addTags(object, model),
+        tags: object.tags,
```
Comment (kubawolanin) [inline bundles/org.openhab.ui.homebuilder/web/src/restItems.js]: What if you add an object or a room by yourself?
You can type your own room name and it'll appear on the list.

Comment (martinvw) [reply bundles/org.openhab.ui.homebuilder/web/src/restItems.js]: It goes fine now (also no errors on the console) and it should be even finer with the new method above, do you agree?

Comment (martinvw) [reply bundles/org.openhab.ui.homebuilder/web/src/restItems.js]: No its not fine, let me do some more tests

Comment (martinvw) [reply bundles/org.openhab.ui.homebuilder/web/src/restItems.js]: No it was already broken, it seems that the `onChanged` is not called when `onNewTag` is called, see:

http://demo.openhab.org:8080/homebuilder/index.html

Comment (martinvw) [reply bundles/org.openhab.ui.homebuilder/web/src/restItems.js]: I will fix that

Comment (martinvw) [reply bundles/org.openhab.ui.homebuilder/web/src/restItems.js]: FTR: https://github.com/vue-generators/vue-form-generator/issues/536

**Comment by martinvw** [comment]
Working properly with `computed` as suggested in the linked issue looks promising but it will take me some time getting this right.

File: bundles/org.openhab.ui.homebuilder/web/src/fieldMultiselect.vue [hunk 3]
```diff
@@ -1,169 +0,0 @@
-<template>
```
Comment (martinvw) [inline bundles/org.openhab.ui.homebuilder/web/src/fieldMultiselect.vue]: Nevermind I will revert this :-)

This handled the icon support it was kind of a fork

**Comment by martinvw** [comment]
> However when switching "Include tags" to "No" the tags are still generated when I click on ITEMS with "Textual Configuration Files" .

Thanks, I repaired that

I also reintroduced the icon-support for the rooms and objects.

**Comment by wborn** [comment]
Thanks the fix works and it's nice to see the icons again.

It would also be nice if the floor selector would worked similar to the room selector. So you can pick the floors you want and don't always have to start with a Cellar, Ground floor etc. That would make more sense for people who live in an apartment.

**Comment by martinvw** [comment]
> It would also be nice if the floor selector would work similar to the room selector. So you can pick the floors you want and don't always have to start with a Cellar, Ground floor etc. That would make more sense for people who live in an apartment.

Can I do that in a follow up PR :-)

**Comment by wborn** [comment]
Does it also generate a modified package-lock.json whenever you do a Maven build @martinvw? :

```
~/git/openhab/openhab-core/bundles/org.openhab.ui.homebuilder$ git diff
diff --git a/bundles/org.openhab.ui.homebuilder/web/package-lock.json b/bundles/org.openhab.ui.homebuilder/web/package-lock.json
index a4a6a7e..da62f3a 100644
--- a/bundles/org.openhab.ui.homebuilder/web/package-lock.json
+++ b/bundles/org.openhab.ui.homebuilder/web/package-lock.json
@@ -3488,12 +3488,14 @@
         "balanced-match": {
           "version": "1.0.0",
           "bundled": true,
-          "dev": true
+          "dev": true,
+          "optional": true
         },
         "brace-expansion": {
           "version": "1.1.11",
           "bundled": true,
           "dev": true,
+          "optional": true,
           "requires": {
             "balanced-match": "^1.0.0",
             "concat-map": "0.0.1"
@@ -3513,7 +3515,8 @@
         "concat-map": {
           "version": "0.0.1",
           "bundled": true,
-          "dev": true
+          "dev": true,
+          "optional": true
         },
         "console-control-strings": {
           "version": "1.1.0",
@@ -3661,6 +3664,7 @@
           "version": "3.0.4",
           "bundled": true,
           "dev": true,
+          "optional": true,
           "requires": {
             "brace-expansion": "^1.1.7"
           }
```

Maybe it's related to these warnings that show during the build:

```
[WARNING] npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):
[WARNING] npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
```

File: bundles/org.openhab.ui.homebuilder/web/package.json [hunk 4]
```diff
@@ -12,16 +12,19 @@
   },
   "dependencies": {
     "ascii-table": "0.0.9",
-    "bootstrap-vue": "^0.18.0",
-    "daemonite-material": "^4.0.0-alpha.6",
+    "bootstrap-vue": "~0.18.0",
+    "daemonite-material": "~4.1.1",
```
Comment (martinvw) [inline bundles/org.openhab.ui.homebuilder/web/package.json]: Because of this upgrade we lost some support for 1920 & > resolutions, so it renders smaller then it used to, is that an issue?

Comment (martinvw) [reply bundles/org.openhab.ui.homebuilder/web/package.json]: It was automatically upgraded because of the ^ but decided to fix it on minors of the current version so that we only update it when we intent to do so.

I was in the process of checking out why it was removed when the rest of the family returned from the swimming pool, now I’m outside supervising the smallest one ☺️

Comment (martinvw) [reply bundles/org.openhab.ui.homebuilder/web/package.json]: If no one has real problems with this I prefer to stick with this version, it’s not a bug it’s a feature, see https://github.com/Daemonite/material/issues/161

So if we want to upgrade any time we would face it again, so better just follow the standards now. It scales good on mobile and smaller screens, only for bigger screens it might be a little step down.

**Comment by martinvw** [comment]
My mvn cleans do not generate a new version, but I just took this build config from the Paper UI build step.

**Comment by wborn** [comment]
> My mvn cleans do not generate a new version

Does it still work for you if you apply my diff? Or will it generate modifications for you when you rebuild with my diff applied? 

It would be rather annoying if these modifications are always generated whenever someone builds openhab-core.

**Comment by martinvw** [comment]
> Does it still work for you if you apply my diff? Or will it generate modifications for you when you rebuild with my diff applied?

No it rolls it back :-(

What happens when you build the paper ui in smarthome?

**Comment by martinvw** [comment]
@wborn https://github.com/npm/npm/issues/20934

**Comment by wborn** [comment]
> I just pushed it to this branch, was not as hard as feared

That's awesome! It seems to work well too! :+1: 

> What happens when you build the paper ui in smarthome?

No files are modified after building it.

> npm/npm#20934

When I update the POM to the following it no longer results in the modifications:

```
          <execution>
            <id>npm install</id>
            <goals>
              <goal>npm</goal>
            </goals>
            <configuration>
              <arguments>install --no-save</arguments>
            </configuration>
          </execution>
```

**Comment by martinvw** [comment]
@wborn no changes for me as well in paper UI but they are cheating the package-lock is not versioned, although the docs explicitly mention that it should be versioned. Maybe @htreu knows why?

I added your suggested fix, all fine now?

File: bundles/org.openhab.ui.homebuilder/pom.xml [hunk 5]
```diff
@@ -73,7 +73,7 @@
           <execution>
             <id>npm install</id>
             <goals>
-              <goal>npm</goal>
+              <goal>npm --no-save</goal>
```
Comment (wborn) [inline bundles/org.openhab.ui.homebuilder/pom.xml]: The goal should remain the same. Just add the configuration like I did. :wink: or it will error:

`[ERROR] Could not find goal 'npm --no-save' in plugin com.github.eirslett:frontend-maven-plugin:1.0 among available goals grunt, install-node-and-npm, jspm, bower, karma, webpack, gulp, npm, ember -> [Help `]`

Comment (martinvw) [reply bundles/org.openhab.ui.homebuilder/pom.xml]: Wout was trying to help so my focus was not that high, changing now and testing locally before pushing :-)

**Comment by martinvw** [comment]
@wborn should be all fine now (I hope) 😆

File: bundles/org.openhab.ui.homebuilder/pom.xml (hunk 5 above)
Comment (wborn) [reply bundles/org.openhab.ui.homebuilder/pom.xml]: You can't start helping with openHAB young enough. ;-)

**Comment by kaikreuzer** [comment]
Guys, I have just analysed why M6 is 25 MB larger than the M5 distro - and I came down to homebuilder, where the jar is now a whopping 25.5 MB (instead of 302KB of before). Extracting the jar, I find 65 MB of node modules packaged with it - not funny!

**Comment by martinvw** [comment]
Is this stopping the release of M6?

If so we could maybe exclude it as standard package? I can  take a look in the train but it’s a short drive today.

**Comment by martinvw** [comment]
> Extracting the jar, I find 65 MB of node modules packaged with it - not funny!

This is btw not because of any new dependencies, before the build was not dynamic but done once on the machine of one developer, I made it part of the maven build, but did not realize the huge increase in size, sorry.

**Comment by kaikreuzer** [comment]
> Is this stopping the release of M6?

No, because I only noticed it AFTER the release 😆.

>  I can take a look in the train but it’s a short drive today.

Not urgent, it is enough to address it in the next days (and before the 2.4 release)!

**Comment by martinvw** [comment]
My jar is already small again, let me test it before creating a PR but it looks easy, I'll keep you posted.
